### PR TITLE
feature: snapshot revert support

### PR DESCRIPTION
**Pull request by ireznice**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,4 +4,5 @@ config/static_machine.yml
 /tmp
 /db/*.sqlite3
 /REMEMBER.md
-storage/*
\ No newline at end of file
+storage/*
```
Comment (lksv) [inline .gitignore]: what is `storage` directory is used for?

File: lib/lab_manager/app/endpoints/computes.rb [hunk 2]
```diff
@@ -132,11 +132,22 @@ class Compute < Base
 
         get '/:id/snapshots/:snapshot_id' do
           compute = ::Compute.find(params[:id])
-          action = compute.snapshots.find(params[:snapshot_id])
-          action.to_json
+          snapshot = compute.snapshots.find(params[:snapshot_id])
```
Comment (lksv) [inline lib/lab_manager/app/endpoints/computes.rb:135]: Snapshot model needs to implement something like `reload_snapshot_data` and needs to be called unless ['false', 'f', '0'].include?(params[:cached])

...similar to GET /compute/:id

File: lib/lab_manager/app/endpoints/computes.rb [hunk 3]
```diff
@@ -132,11 +132,22 @@ class Compute < Base
...
+          snapshot.to_json
         end
 
         get '/:id/snapshots/:snapshot_id/revert' do
```
Comment (lksv) [inline lib/lab_manager/app/endpoints/computes.rb]: It has to be POST method

File: lib/lab_manager/app/endpoints/computes.rb [hunk 4]
```diff
@@ -132,11 +132,22 @@ class Compute < Base
...
+          compute = ::Compute.find(params[:id])
+          snapshot = compute.snapshots.find(params[:snapshot_id])
+
+          action = compute.actions.create!(
+            command: 'revert_snapshot_vm',
+            payload: {
+              snapshot_id: snapshot.id,
+            }
+          )
```
Comment (lksv) [inline lib/lab_manager/app/endpoints/computes.rb:148]: For a nice design wouldn't we encapsulate this logic somewhere? We are reverting snapshot, but actually we are creating an action on compute. It's breaking several design patters. (The same counts for create snapshot.)
We can discuss personally about it on Monday.

File: lib/lab_manager/workers/action_worker.rb [hunk 5]
```diff
@@ -182,6 +183,26 @@ def take_snapshot_vm
...
+    def revert_snapshot_vm
+      lock { compute.revert! }
+      begin
+        fail 'Wrong action payload' unless action.payload
+        fail 'Wrong action payload, no snapshot_id provided' unless action.payload.has_key?(:snapshot_id)
+        snapshot = compute.snapshots.find(action.payload[:snapshot_id])
+        # we suppose, that revert process should be fully finished before performing
+        # some actions such as restart, shutdown, reboot, power_off etc.
+        lock do
+          compute.revert_snapshot_vm( { name: snapshot.name } )
```
Comment (lksv) [inline lib/lab_manager/workers/action_worker.rb:195]: what happens when we have got snapshot form stopped machine and revert to it when machine is running?
Shouldn't we have the current state and set the compute model state accordingly?

File: spec/integration/endpoints/computes_spec.rb [hunk 6]
```diff
@@ -319,5 +319,48 @@ def app
       end
     end
 
+    describe 'GET /computes/:compute_id/snaphosts/:snapshot_id/revert' do
```
Comment (lksv) [inline spec/integration/endpoints/computes_spec.rb]: it has to be POST method

File: .gitignore (hunk 1 above)
Comment (ireznice) [inline .gitignore]: add /

File: lib/lab_manager/models/compute.rb [hunk 7]
```diff
@@ -95,7 +96,9 @@ class CannotDeleteAliveVM < RuntimeError
     event :resumed       do transitions from: :resuming,       to: :running       end
 
     event :revert        do transitions from: :running,        to: :reverting     end
-    event :reverted      do transitions from: :reverting,      to: :running       end
+    event :reverted_r    do transitions from: :reverting,      to: :running       end
```
Comment (lksv) [inline lib/lab_manager/models/compute.rb]: revert_running

File: lib/lab_manager/models/compute.rb [hunk 8]
```diff
@@ -95,7 +96,9 @@ class CannotDeleteAliveVM < RuntimeError
...
+    event :reverted_o    do transitions from: :reverting,      to: :powered_off   end
```
Comment (lksv) [inline lib/lab_manager/models/compute.rb]: revert_power_off

File: lib/lab_manager/provider/vsphere.rb [hunk 9]
```diff
@@ -318,6 +318,37 @@ def take_snapshot_vm(opts)
...
+        Retryable.retryable(tries: 3, exception_cb: RETRYABLE_CALLBACK) do
+          server = vs.servers.get(instance_uuid)
+        end
+
+        Retryable.retryable(tries: 3, exception_cb: RETRYABLE_CALLBACK) do
+          snapshot = server.snapshots.all(recursive: true).find do |t|
+            t.name == opts[:name]
+          end
+        end
+        p snapshot.inspect
```
Comment (lksv) [inline lib/lab_manager/provider/vsphere.rb]: please remove

File: lib/lab_manager/provider/vsphere.rb [hunk 10]
```diff
@@ -361,6 +384,13 @@ def set_provider_data(vm_instance_data = nil, full: false, vs: nil)
       end
     end
 
+    def vm_state
+      set_provider_data
+      compute.provider_data['power_state'] || 'PoweredOff'
+    rescue
+      'PoweredOff'
```
Comment (lksv) [inline lib/lab_manager/provider/vsphere.rb]: `:power_off`

File: lib/lab_manager/provider/vsphere.rb [hunk 11]
```diff
@@ -361,6 +384,13 @@ def set_provider_data(vm_instance_data = nil, full: false, vs: nil)
       end
     end
 
+    def vm_state
+      set_provider_data
+      compute.provider_data['power_state'] || 'PoweredOff'
```
Comment (lksv) [inline lib/lab_manager/provider/vsphere.rb]: should be one of `:power_on` `:power_off`, 
...any other?

File: lib/lab_manager/workers/action_worker.rb [hunk 12]
```diff
@@ -182,6 +183,33 @@ def take_snapshot_vm
       action.save!
     end
 
+    def revert_snapshot_vm
+      lock { compute.revert! }
+      begin
+        fail 'Wrong action payload' unless action.payload
```
Comment (lksv) [inline lib/lab_manager/workers/action_worker.rb]: ArgumentError

File: spec/integration/endpoints/computes_spec.rb [hunk 13]
```diff
@@ -319,5 +319,48 @@ def app
...
+          post "/computes/#{compute.id}/snapshots/#{snapshot.id}/revert"
+        end
+      end
+
+      it 'schedules sidekiq job for action', sidekiq: true do
+        expect do
+          post "/computes/#{compute.id}/snapshots/#{snapshot.id}/revert"
+          expect(last_response.status).to eq 200
+        end.to change(LabManager::ActionWorker.jobs, :size).by(1)
+      end
```
Comment (lksv) [inline spec/integration/endpoints/computes_spec.rb:363]: please add test for revert from running machine to stopped machine and visa versa.

File: spec/unit/lab_manager/action_worker_spec.rb [hunk 14]
```diff
@@ -135,7 +136,7 @@
       action_worker.perform(action.id)
       snapshot.reload
       expect(snapshot.provider_data.symbolize_keys).to eq sample_provider_data
-    end
+      end
```
Comment (lksv) [inline spec/unit/lab_manager/action_worker_spec.rb]: mistype?

File: spec/integration/endpoints/computes_spec.rb (hunk 13 above)
Comment (ireznice) [inline spec/integration/endpoints/computes_spec.rb:363]: surely not in this file

**Comment by ireznice** [comment]
@lksv fixed, please check it out again

**Comment by ireznice** [comment]
@lksv fixed, please check it out

**Comment by ireznice** [comment]
@lksv  fixed, please check it out

**Pull request merged**